app/scripts/modules/core/src/artifact/react/ExpectedArtifactEditor.tsx

```diff
@@ -0,0 +1,144 @@
+      account: this.state.account,
+    });
+  };
+
+  private accountsForExpectedArtifact(expectedArtifact: IExpectedArtifact): IArtifactAccount[] {
+    const artifact = ExpectedArtifactService.artifactFromExpected(expectedArtifact);
+    if (!artifact) {
+      return [];
+    }
+    return this.props.accounts.filter(a => a.types.includes(artifact.type));
+  }
+
+  private onSourceChange = (e: IExpectedArtifactSourceOption) => {
+    this.setState({ source: e });
+  };
+
+  private onKindChange = (kind: IArtifactKindConfig) => {
+    const expectedArtifact = cloneDeep(this.state.expectedArtifact);
+    expectedArtifact.matchArtifact.type = kind.type;
+    expectedArtifact.matchArtifact.kind = kind.key;
+    const accounts = this.accountsForExpectedArtifact(expectedArtifact);
+    this.setState({ expectedArtifact, account: accounts[0] });
+  };
+
+  private onAccountChange = (account: IArtifactAccount) => {
+    this.setState({ account });
+  };
+
+  private onArtifactEdit = (artifact: IArtifact) => {
+    const expectedArtifact = { ...this.state.expectedArtifact, matchArtifact: { ...artifact } };
+    let account = this.state.account;
+    if (this.state.expectedArtifact.matchArtifact.type !== artifact.type) {
+      const accounts = this.accountsForExpectedArtifact(expectedArtifact);
+      account = accounts[0];
+    }
+    this.setState({ expectedArtifact, account });
+  };
+
+  private availableKinds = () => {
+    const { kinds, accounts } = this.props;
+    return kinds.filter(k => k.key === 'custom' || accounts.find(a => a.types.includes(k.type)));
+  };
+
+  public render() {
+    const { sources } = this.props;
+    const { expectedArtifact, source, account } = this.state;
+    const accounts = this.accountsForExpectedArtifact(expectedArtifact);
+    const artifact = ExpectedArtifactService.artifactFromExpected(expectedArtifact);
+    const kinds = this.availableKinds().sort((a, b) => a.label.localeCompare(b.label));
+    const kind = artifact ? kinds.find(k => k.key === artifact.kind) : null;
+    const EditCmp = kind && kind.editCmp;
+    return (
+      <div>
+        <StageConfigField label="Artifact Source" fieldColumns={8}>
+          <ExpectedArtifactSourceSelector sources={sources} selected={source} onChange={this.onSourceChange} />
+        </StageConfigField>
+        <StageConfigField label="Artifact Type" fieldColumns={8}>
+          <ExpectedArtifactKindSelector kinds={kinds} selected={kind} onChange={this.onKindChange} />
+        </StageConfigField>
+        <StageConfigField label="Artifact Account" fieldColumns={8}>
+          <ArtifactAccountSelector accounts={accounts} selected={account} onChange={this.onAccountChange} />
+        </StageConfigField>
+        {EditCmp && <EditCmp artifact={artifact} onChange={this.onArtifactEdit} labelColumns={3} fieldColumns={3} />}
+        <StageConfigField label="" fieldColumns={8}>
+          <button onClick={this.onSave}>Create New Artifact</button>
```

Comment:
Would add standard spinnaker styles to this button. Probably `add-new btn btn-block btn-sm`.

Reply:
I'm wondering if the language is a little confusing - since inside your dropdown you have "Create new", and then here you have "Create New Artifact". What about "Save" or "Confirm"?

Reply:
Agreed. I thought about this while I was writing it and figured "Save" would clash in meaning with the "Save Changes" button for the pipeline.  I'll use "Confirm".

Reply:
I ended up using `btn-primary` instead of `add-new` because the dotted outline from `add-new` gives the sense that it's going to add something / fill a gap in the form rather than commit the existing changes:

<img width="663" alt="screen shot 2018-09-05 at 11 04 33 am" src="https://user-images.githubusercontent.com/34253460/45102477-d4e72c80-b0fb-11e8-81f6-76563da2809d.png">